java-manta-client/src/main/java/com/joyent/manta/client/crypto/EncryptingPartEntity.java

```diff
@@ -0,0 +1,119 @@
+package com.joyent.manta.client.crypto;
+
+import com.joyent.manta.client.multipart.MultipartOutputStream;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHeaders;
+import org.apache.http.entity.ContentType;
+import org.apache.http.message.BasicHeader;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Class that provides an interface for writing multipart parts to a
+ * HTTP socket represented as a {@link OutputStream}.
+ */
+public class EncryptingPartEntity implements HttpEntity {
+    /**
+     * Value for an unknown stream length.
+     */
+    public static final long UNKNOWN_LENGTH = -1L;
+
+    /**
+     * Content transfer encoding to send (set to null for none).
+     */
+    private static final Header CRYPTO_TRANSFER_ENCODING = null;
+
+    /**
+     * Content type to store encrypted content as.
+     */
+    private static final Header CRYPTO_CONTENT_TYPE = new BasicHeader(
+            HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_OCTET_STREAM.toString());
+
+    /**
+     * Encrypting stream.
+     */
+    private final OutputStream cipherStream;
+
+    /**
+     * Underlying entity that is being encrypted.
+     */
+    private final HttpEntity wrapped;
+
+    /**
+     * Multipart stream that allows to attaching / detaching streams.
+     */
+    private final MultipartOutputStream multipartStream;
+
+    /**
+     * Creates a new instance based on the specified parameters.
+     *
+     * @param cipherStream encrypting stream
+     * @param multipartStream multipart stream that allows to attaching / detaching streams
+     * @param wrapped wrapped entity to encrypt and send
+     */
+    public EncryptingPartEntity(final OutputStream cipherStream,
+                                final MultipartOutputStream multipartStream,
+                                final HttpEntity wrapped) {
+        this.cipherStream = cipherStream;
+        this.multipartStream = multipartStream;
+        this.wrapped = wrapped;
+    }
+
+    @Override
+    public boolean isRepeatable() {
+        return wrapped.isRepeatable();
+    }
+
+    @Override
+    public boolean isChunked() {
+        return true;
+    }
+
+    @Override
+    public long getContentLength() {
+        // if content length is not exact, then we will hang on a socket timeout
+        return UNKNOWN_LENGTH;
```

Comment:
Actually, allowing us to calculate an accurate content-length will increase perf by about 10-30% because it means that we don't have to transmit in chunked mode.